### PR TITLE
Feature/tx mst handle

### PR DESCRIPTION
**Pull request by l4l**
In this PR TransactionProcessorImpl is able to handle multisignature transaction.

Some refactoring was done in c52c95c, there should be no (hopefully) logic changes. List of refactorings:
- Most of the types was moved out from mst_types to model ones
- Mst is fixed for using it's own type `DataType`, that basically an alias for `std::shared_ptr<iroha::model::Transaction>`
- In mst state is made an explicit copy of tx
- [tx_processor](https://github.com/hyperledger/iroha/pull/690/commits/c52c95cb42ecdd04e3f5bd3cd359e2f4aa8942b7#diff-2555aa1c75c15355ee489fce2274d681) code was shortened of boost (`join`, `for_each`, `adaptor::filtered`)
- also in [tx_processor](https://github.com/hyperledger/iroha/pull/690/commits/c52c95cb42ecdd04e3f5bd3cd359e2f4aa8942b7#diff-2555aa1c75c15355ee489fce2274d681) `TransactionResponse` is inlined, where it's possible

This also introduces a check of tx' [signatures](https://github.com/hyperledger/iroha/pull/690/files#diff-2555aa1c75c15355ee489fce2274d681R110), that also a good point for discussion

Note that it doesn't built due issue related FairMstProcessor problems usage (that's because of MstStorageStateImpl because of Peer, if you care). If you'd like to test you may build&run following targets:
- transaction_processor_test
- mst_processor_test
- maybe some other that I missed

### CODE REVIEW

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 1]
```diff
@@ -15,103 +15,118 @@
...
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status = TransactionResponse::COMMITTED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              candidate_set_.clear();
+              boost::for_each(
+                  boost::join(this->proposal_set_, this->candidate_set_),
+                  notify_fail);
```
Comment (muratovv) [inline irohad/torii/processor/impl/transaction_processor_impl.cpp]: Notify fail provides only in the end observable. This is wrong behavior at all.

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 2]
```diff
@@ -15,103 +15,118 @@
...
+        response.current_status = Status::STATELESS_VALIDATION_SUCCESS;
+        switch (transaction->signatures.size()) {
+          case 0:
+            response.current_status = Status::STATELESS_VALIDATION_FAILED;
+            break;
+          case 1:
+            pcs_->propagate_transaction(transaction);
+            break;
+          default:
+            mst_proc_->propagateTransaction(transaction);
```
Comment (muratovv) [inline irohad/torii/processor/impl/transaction_processor_impl.cpp]: `propagateTransaction` not so suitable name. Maybe refactor it?

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 3]
```diff
@@ -15,103 +15,118 @@
...
       log_->info("handle transaction");
-      model::TransactionResponse response;
-      response.tx_hash = hash(*transaction).to_string();
-      response.current_status =
-          model::TransactionResponse::Status::STATELESS_VALIDATION_FAILED;
+      TransactionResponse response{hash(*transaction).to_string(),
+                                   Status::STATELESS_VALIDATION_FAILED};
 
+      // TODO: nice place for code linearizing
       if (validator_->validate(*transaction)) {
```
Comment (muratovv) [inline irohad/torii/processor/impl/transaction_processor_impl.cpp]: Also, feel that providing information about MST transactions it is not a responsibility of validator. Need to move this business logic to tx processor.

File: irohad/multi_sig_transactions/mst_types.hpp [hunk 4]
```diff
@@ -18,23 +18,13 @@
...
-  using ConstRefT = const T &;
-
-  using ConstRefTransaction = ConstRefT<TransactionType>;
-  using ConstRefPeer = ConstRefT<iroha::model::Peer>;
-  using ConstRefTime = ConstRefT<TimeType>;
   using ConstRefState = ConstRefT<MstState>;
-} // namespace iroha
-#endif //IROHA_MST_TYPES_HPP
+
+  using DataType = SharedTx;
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/multi_sig_transactions/mst_types.hpp]: I also think so, but want to leave this refactoring for future work as long as mst uses a lot `DataType`

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 5]
```diff
@@ -15,103 +15,118 @@
...
+        std::shared_ptr<StatelessValidator> validator,
+        std::shared_ptr<MstProcessor> mst_proc)
+        : pcs_(std::move(pcs)),
+          validator_(std::move(validator)),
+          mst_proc_(std::move(mst_proc)) {
       log_ = logger::log("TxProcessor");
 
       // insert all txs from proposal to proposal set
       pcs_->on_proposal().subscribe([this](model::Proposal proposal) {
         for (const auto &tx : proposal.transactions) {
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: tried, looks horrible

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 6]
```diff
@@ -15,103 +15,118 @@
...
-          response.current_status =
-              TransactionResponse::STATELESS_VALIDATION_SUCCESS;
           notifier_.get_subscriber().on_next(
-              std::make_shared<model::TransactionResponse>(response));
+              std::make_shared<TransactionResponse>(TransactionResponse{
+                  hash(tx).to_string(), Status::STATELESS_VALIDATION_SUCCESS}));
         }
       });
 
+      static const auto notify_success = [this](const auto &tx) {
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: 1. Function initialized only once
2. No need bind in lambdas

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 7]
```diff
@@ -15,103 +15,118 @@
...
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto block) {
+              static const auto in_proposal = [this](const auto &tx) {
+                return this->proposal_set_.count(hash(tx).to_string());
+              };
+              boost::for_each(
+                  block.transactions | boost::adaptors::filtered(in_proposal),
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: It's boost pipe operator

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 1 above)
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: It's on complete callback, what's wrong?

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 2 above)
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: Imo we should refactor all the torii-related stuff in separate pr, it will be quite big

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: Meh, that's UB actually, static object capture `this`, thanks

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 2 above)
Comment (muratovv) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: I mean `propagateTransaction` from MST.

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 8]
```diff
@@ -15,106 +15,123 @@
...
+              boost::for_each(this->candidate_set_, [this](auto &t) {
+                return this->notify_commit(t);
+              });
+              this->candidate_set_.clear();
             });
       });
+
+      mst_proc_->onPreparedTransactions().subscribe(
+          [this](auto tx) { return this->transactionHandle(tx); });
+      mst_proc_->onExpiredTransactions().subscribe(
```
Comment (lebdron) [inline irohad/torii/processor/impl/transaction_processor_impl.cpp]: `STATEFUL_VALIDATION_FAILED` does not seem like an appropriate response for this case, since the transaction did not pass through `Simulator` yet. `STATELESS_VALIDATION_FAILED` seems better to me, because the transaction just failed to accumulate enough signatures, whether the signatures correspond to creator or not.

File: irohad/main/application.cpp [hunk 9]
```diff
@@ -213,9 +214,22 @@ void Irohad::initPeerCommunicationService() {
   log_->info("[Init] => pcs");
 }
 
+void Irohad::initMstProcessor() {
+  auto mst_transport = std::make_shared<MstTransportGrpc>();
+  auto mst_completer = std::make_shared<DefaultCompleter>();
+  auto mst_storage = std::make_shared<MstStorageStateImpl>(mst_completer);
+  // TODO: @l4l magics should be fixed with options in cli branch
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/main/application.cpp]: I tried to change files as minimal as possible, even wrong formatting may left
I hope that can help in merging

File: irohad/main/application.cpp [hunk 10]
```diff
@@ -213,9 +214,22 @@ void Irohad::initPeerCommunicationService() {
...
+  auto mst_propagation = std::make_shared<GossipPropagationStrategy>(
+      wsv, std::chrono::milliseconds(5) /*emitting period*/,
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/main/application.cpp]: Nope, checked again. Maybe due to mst-trunk has outdated clang-format?

File: irohad/model/transaction_response.hpp [hunk 11]
```diff
@@ -34,18 +34,23 @@ namespace iroha {
...
+        /// transaction is received, but not validated
+        ON_PROCESS,
+        /// transaction is not in handler map
+        NOT_RECEIVED
       };
 
-      Status current_status;
-
-      virtual ~TransactionResponse() = default;
+      Status current_status{};
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/model/transaction_response.hpp]: It's enum :( I don't really sure I meant to add this braces actually

File: irohad/multi_sig_transactions/mst_time_provider_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <chrono>
+#include "multi_sig_transactions/mst_time_provider.hpp"
+
+namespace iroha {
+
+  class MstTimeProviderImpl : public MstTimeProvider {
+   public:
+    TimeType getCurrentTime() const override {
+      return std::chrono::system_clock::now().time_since_epoch().count();
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/multi_sig_transactions/mst_time_provider_impl.hpp]: Afaik it's already updated. Thanks for an advice

File: irohad/multi_sig_transactions/state/impl/mst_state.cpp [hunk 13]
```diff
@@ -103,14 +101,14 @@ namespace iroha {
       return;
     }
 
-    // state already contains transaction, merge signatures
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/multi_sig_transactions/state/impl/mst_state.cpp]: I though `merge_unique` is mean exactly this

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 14]
```diff
@@ -15,106 +15,123 @@
  * limitations under the License.
  */
 
-#include <crypto/hash.hpp>
-#include <endpoint.pb.h>
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/algorithm/for_each.hpp>
+#include <boost/range/join.hpp>
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: It was used in 25b9923 actually, but yeah, now redudant

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 15]
```diff
@@ -15,106 +15,123 @@
...
       pcs_->on_proposal().subscribe([this](model::Proposal proposal) {
         for (const auto &tx : proposal.transactions) {
           proposal_set_.insert(hash(tx).to_string());
-          TransactionResponse response;
-          response.tx_hash = hash(tx).to_string();
-          response.current_status =
-              TransactionResponse::STATELESS_VALIDATION_SUCCESS;
           notifier_.get_subscriber().on_next(
-              std::make_shared<model::TransactionResponse>(response));
+              std::make_shared<TransactionResponse>(TransactionResponse{
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: It has no constructor actually, that why it done this way

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 16]
```diff
@@ -15,106 +15,123 @@
...
-                  candidate_set_.insert(hash(tx).to_string());
-                  TransactionResponse response;
-                  response.tx_hash = hash(tx).to_string();
-                  response.current_status =
-                      model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS;
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto block) {
```
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: I'll better change it to universal reference, thanks

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 8 above)
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: Agree, maybe introduce `MST_VALIDATION_FALED`? For me it seems pretty useful to distinguish this errors

File: test/module/irohad/torii/processor/transaction_processor_test.cpp [hunk 17]
```diff
@@ -84,22 +93,88 @@ TEST_F(TransactionProcessorTest,
...
+  tp->transactionHandle(tx);
+  mst_prepared_notifier.get_subscriber().on_next(tx);
+
+  ASSERT_TRUE(wrapper.validate());
+}
+
+/**
+ * @given multisig tx
+ * @when propagate it with big quorum
+ * @then ensure after expiring it leads to STATEFUL_VALIDATION_FAILED
```
[the remark this answers is not in the document]
Comment (l4l) [reply test/module/irohad/torii/processor/transaction_processor_test.cpp]: [Here's](https://github.com/hyperledger/iroha/pull/690#discussion_r151048148) the reason

File: irohad/main/application.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (lebdron) [reply irohad/main/application.cpp]: >outdated clang-format

Yes, missed that point.

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 8 above)
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: Agreed on the new `EXPIRED` status

File: irohad/main/application.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/main/application.cpp]: This is actually related to the #661 and that task in jira seems deleted

File: irohad/multi_sig_transactions/state/impl/mst_state.cpp [hunk 18]
```diff
@@ -103,14 +101,14 @@ namespace iroha {
...
-    (*corresponding)->signatures = merge_unique<iroha::model::SignatureHasher>(
-        (*corresponding)->signatures, rhs_tx->signatures);
+    auto tx = *corresponding;
```
[the remark this answers is not in the document]
Comment (lebdron) [reply irohad/multi_sig_transactions/state/impl/mst_state.cpp]: Set elements are pointers. Pointers themselves don't change, no undefined behavior here.

File: irohad/main/application.cpp [hunk 19]
```diff
@@ -213,9 +214,23 @@ void Irohad::initPeerCommunicationService() {
...
 }
 
+void Irohad::initMstProcessor() {
+  auto mst_transport = std::make_shared<MstTransportGrpc>();
+  auto mst_completer = std::make_shared<DefaultCompleter>();
+  auto mst_storage = std::make_shared<MstStorageStateImpl>(mst_completer);
+  // TODO: @l4l magics should be fixed with options in cli branch
+  //            check #661 for details
+  auto mst_propagation = std::make_shared<GossipPropagationStrategy>(
+      wsv, std::chrono::milliseconds(5) /*emitting period*/,
```
Comment (muratovv) [inline irohad/main/application.cpp]: Every five milliseconds propagate state to five peers. I guess default preferences should be like 5 seconds and 2 peers. But this parameter should be set by the runner.

File: irohad/multi_sig_transactions/mst_time_provider_impl.hpp [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_MST_TIME_PROVIDER_IMPL_HPP
+#define IROHA_MST_TIME_PROVIDER_IMPL_HPP
+
+#include <chrono>
```
Comment (muratovv) [inline irohad/multi_sig_transactions/mst_time_provider_impl.hpp]: I guess `chrono` include is needless here.

File: irohad/multi_sig_transactions/storage/impl/mst_storage.cpp [hunk 21]
```diff
@@ -20,21 +20,20 @@
...
 
 namespace iroha {
-  MstStorage::MstStorage() {
-    log_ = logger::log("MstStorage");
-  }
+  MstStorage::MstStorage() { log_ = logger::log("MstStorage"); }
 
-  MstState MstStorage::apply(ConstPeer &target_peer, const MstState &new_state) {
+  MstState MstStorage::apply(ConstPeer &target_peer,
+                             const MstState &new_state) {
```
[the remark this answers is not in the document]
Comment (muratovv) [reply irohad/multi_sig_transactions/storage/impl/mst_storage.cpp]: @l4l This creates inconsistency in API. If you assume that types should be explicitly written using this approach for both cases.

File: irohad/torii/processor/impl/transaction_processor_impl.cpp [hunk 22]
```diff
@@ -15,106 +15,109 @@
...
-                  candidate_set_.insert(hash(tx).to_string());
-                  TransactionResponse response;
-                  response.tx_hash = hash(tx).to_string();
-                  response.current_status =
-                      model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS;
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto &&block) {
```
Comment (muratovv) [inline irohad/torii/processor/impl/transaction_processor_impl.cpp]: I don't understand what these changes do it integration MST PR. I guess more confident to make it in separate PR.

File: test/module/irohad/torii/processor/transaction_processor_test.cpp [hunk 23]
```diff
@@ -29,52 +31,60 @@ using namespace iroha::torii;
...
+  std::shared_ptr<MockMstProcessor> mp;
 };
 
 /**
  * Transaction processor test case, when handling stateless valid transaction
  */
-TEST_F(TransactionProcessorTest,
-     TransactionProcessorWhereInvokeValidTransaction) {
-
+TEST_F(TransactionProcessorTest, ValidTransaction) {
```
Comment (muratovv) [inline test/module/irohad/torii/processor/transaction_processor_test.cpp]: Pls, also, rework documentation string with `given`, `when`, `then` approach.

File: test/module/irohad/torii/processor/transaction_processor_test.cpp [hunk 24]
```diff
@@ -84,22 +94,88 @@ TEST_F(TransactionProcessorTest,
 /**
  * Transaction processor test case, when handling invalid transaction
  */
-TEST_F(TransactionProcessorTest,
-     TransactionProcessorWhereInvokeInvalidTransaction) {
-
+TEST_F(TransactionProcessorTest, InvalidTransaction) {
```
Comment (muratovv) [inline test/module/irohad/torii/processor/transaction_processor_test.cpp]: Same about documentation.

File: schema/endpoint.proto [hunk 25]
```diff
@@ -16,6 +16,7 @@ enum TxStatus {
   COMMITTED = 4;
   ON_PROCESS = 5;
   NOT_RECEIVED = 6;
+  EXPIRED = 7;
```
Comment (muratovv) [inline schema/endpoint.proto]: Is there an available conversion from proto to model and vice versa?

File: irohad/main/application.cpp (hunk 19 above)
Comment (l4l) [reply irohad/main/application.cpp]: Upon merging cli-options that should be fixed, for now it's just some random values. I can fix exact values if it matters

File: schema/endpoint.proto (hunk 25 above)
Comment (l4l) [reply schema/endpoint.proto]: Check 612cbe6, and I didn't find reverse way convertion

File: irohad/torii/processor/impl/transaction_processor_impl.cpp (hunk 22 above)
Comment (l4l) [reply irohad/torii/processor/impl/transaction_processor_impl.cpp]: Sorry, for that inconvenience
Don't really want to copy-paste similar code that's why this refactoring is made (though boost is not that necessary)
